src/test/test_pidfile.sh

```diff
@@ -61,15 +61,41 @@ function teardown_unexist_pidfile() {
     return $status
 }
 
-function TEST_contend_pidfile() {
+function TEST_pidfile() {
     local dir=$1
     setup $dir 
- 
-    run_mon $dir a 
+
+    # no daemon can use a pidfile that is owned by another daemon
+    run_mon $dir a || return 1
     run_mon $dir a 2>&1 | grep "failed to lock pidfile" || return 1
 
-    run_osd $dir 0 
+    run_osd $dir 0 || return 1
     run_osd $dir 0 2>&1 | grep "failed to lock pidfile" || return 1
+
+    # when a daemon shutdown, it will not unlink a path different from
+    # the one it owns
+    mv $dir/osd.0.pid $dir/osd.0.pid.old || return 1
+    cp $dir/osd.0.pid.old $dir/osd.0.pid || return 1
+    kill_daemons $dir TERM osd.0 || return 1
+    test -f $dir/osd.0.pid || return 1
+
+    # when a daemon starts, it re-uses the pid file if no other daemon
+    # has it locked
+    run_osd $dir 0 || return 1
+    ! cmp $dir/osd.0.pid $dir/osd.0.pid.old || return 1
+
+    # if the pid in the file is different from the pid of the daemon
+    # the file is not removed because it is assumed to be owned by
+    # another daemon
+    echo 123 > $dir/osd.0.pid
+    kill_daemons $dir TERM osd.0 || return 1
```

Comment:
@dachary  kill_daemons at line 90 leads to osd daemon not killed at final, so  a cp $dir/osd.0.pid $dir/osd.0.pid.old is needed before changing content in osd.0.pid